spec/ics-3-connection-semantics/README.md

```diff
@@ -0,0 +1,158 @@
+
+A *actor*, as referred to in this specification, is an entity capable of executing datagrams who is paying for computation / storage (via gas or a similar mechanism) but is otherwise untrusted. Possible actors include:
+- End users signing with an account key
+- On-chain smart contracts acting autonomously or in response to another transaction
+- On-chain modules acting in response to another transaction or in a scheduled manner
+
+#### Requirements
+
+Connection handlers and subsequent protocols make use of a simple key-value store interface provided by the underlying state machine. This store must provide two functions, which behave in the way you would expect:
+- `Get(Key) -> Value | null`
+- `Set(Key, Value)`
+
+`Key` and `Value` are assumed to be byte slices; encoding details are left to a later ICS.
+
+Blockchains also need the ability to introspect their own root-of-trust (with `getRootOfTrust`) in order to confirm that the connecting chain has stored the correct one.
+
+#### Subprotocols
+
+Subprotocols are defined as a set of datagram types and a `handleDatagram` function which must be implemented by the state machine of the implementing blockchain. Datagrams must be relayed between chains by an external process. This process is assumed to behave in an arbitrary manner — no safety properties are dependent on its behavior, although progress is generally dependent on the existence of at least one correct relayer process. Further discussion is deferred to [ICS 18: Relayer Algorithms](../spec/ics-18-relayer-algorithms).
+
+IBC subprotocols are reasoned about as interactions between two chains `A` and `B` — there is no prior distinction between these two chains and they are assumed to be executing the same, correct IBC protocol. `A` is simply by convention the chain which goes first in the subprotocol and `B` the chain which goes second. Protocol definitions should generally avoid including `A` and `B` in variable names to avoid confusion (as the chains themselves do not know whether they are `A` or `B` in the protocol).
+
+This ICS defines four subprotocols: opening handshake, header tracking, closing handshake, and closing by equivocation.
+
+##### Opening Handshake
+
+The opening handshake subprotocol serves to initialize roots of trust for two chains on each other and negotiate an agreeable connection version.
+
+This subprotocol need not be permissioned, modulo anti-spam measures.
+
+![Opening Handshake](opening_handshake.png)
+
+A correct protocol execution flows as follows:
+
+| Initiator | Datagram          | Chain |
+| --------- | ----------------- | ----- |
+| Actor     | `ConnOpenInit`    | A     |
+| Relayer   | `ConnOpenTry`     | B     |
+| Relayer   | `ConnOpenAck`     | A     |
+| Relayer   | `ConnOpenConfirm` | B     |
+
```

Comment:
The table helps a lot! A state machine diagram would also be nice :) A few questions:
What happens if both sides start the connection in the same time?
What happens if version negotiation fails?
Are or can any of these messages be used after the 4-way handshake is done? How are they handled?
Can the connection be made in a single direction?

Reply:
Can you give an example of the kind of state machine diagram you mean? That sounds useful.

The `asserts` in the protocol description guard against datagram mis-ordering.

I still need to write out timeout logic.

I think we will also have unidirectional connections but I was saving that spec for later since it should be easy to construct once we're agreed on the bidirectional connection semantics.

Reply:
> Can you give an example of the kind of state machine diagram you mean? That sounds useful.

Something like that...not sure I got it right though.

![image](https://user-images.githubusercontent.com/6361427/55418702-7986fb00-5573-11e9-855a-bf82346ac84e.png)

Can be ASCII art :)
Nit: OPENTRY and TRYOPEN are both used in the spec

Reply:
Same for the other (sub)protocols. I can make them if you tell me what to use. And then you can correct them.
BTW, one other thing the spec should probably mention are timeouts, i.e. how long we wait for a datagram and also specify the action taken, e.g. retrying, closing etc.